tests/Microsoft.Identity.Test.Unit/ManagedIdentityTests/ManagedIdentityTests.cs

```diff
@@ -480,5 +483,168 @@ public async Task ManagedIdentityCacheTestAsync()
                 appTokenCacheRecoder.AssertAccessCounts(1, 1);
             }
         }
+
+        [DataTestMethod]
+        [DataRow(0)]
+        [DataRow(1)]
+        [DataRow(2)]
+        [DataRow(3)]
+        public async Task ManagedIdentityRefreshOnIsCalculatedAsync(int expiresInHours)
+        {
+            AuthenticationResult result;
+
+            using (new EnvVariableContext())
+            using (var httpManager = new MockHttpManager())
+            {
+                SetEnvironmentVariables(ManagedIdentitySource.AppService, AppServiceEndpoint);
+
+                var mi = ManagedIdentityApplicationBuilder.Create()
+                    .WithExperimentalFeatures()
+                    .WithHttpManager(httpManager)
+                    .Build();
+
+                httpManager.AddManagedIdentityMockHandler(
+                    AppServiceEndpoint,
+                    Resource,
+                    MockHelpers.GetMsiSuccessfulResponse(expiresInHours),
+                    ManagedIdentitySource.AppService);
+
+                AcquireTokenForManagedIdentityParameterBuilder builder = mi.AcquireTokenForManagedIdentity(Resource);
+
+                if (expiresInHours == 0)
+                {
+                    MsalClientException ex = await AssertException.TaskThrowsAsync<MsalClientException>(
+                        () => builder.ExecuteAsync()).ConfigureAwait(false);
+
+                    Assert.AreEqual(ex.ErrorCode, "invalid_token_provider_response_value");
+                    return;
+                }
+                else
+                {
+                    result = await builder.ExecuteAsync().ConfigureAwait(false);
+                }
+
+                Assert.IsNotNull(result);
+                Assert.IsNotNull(result.AccessToken);
+                Assert.AreEqual(TokenSource.IdentityProvider, result.AuthenticationResultMetadata.TokenSource);
+
+                Assert.AreEqual(ApiEvent.ApiIds.AcquireTokenForSystemAssignedManagedIdentity, builder.CommonParameters.ApiId);
+
+                switch (expiresInHours)
+                {
+                    case 0:
+                    case 1:
+                    case 2:
+                        Assert.IsFalse(result.AuthenticationResultMetadata.RefreshOn.HasValue);
```

Comment:
nit: this could be a separate parameter in the test - `refreshOnHasValue`
```csharp
        [DataRow(0, false)]
        [DataRow(1, false)]
        [DataRow(2, false)]
        [DataRow(3, true)]
        public async Task ManagedIdentityRefreshOnIsCalculatedAsync(int expiresInHours)
```
and then only
```csharp
Assert.AreEqual(refreshOnHasValue, result.AuthenticationResultMetadata.RefreshOn.HasValue);
```